src/controllers/api-key.js

```diff
@@ -10,52 +10,184 @@
+    if (!organizations.includes(imsOrgId)) {
+      throw new ErrorWithStatusCode('Invalid request: Organization not found', STATUS_UNAUTHORIZED);
+    }
   }
 
-  // eslint-disable-next-line no-unused-vars
-  function validateImsOrgId(imsOrgId) {
-    // Not implemented yet
+  /**
+   * Get the IMS user token from the headers.
+   * @param headers
+   * @returns string imsUserToken
+   */
+  function getImsUserToken(headers) {
+    const authorizationHeader = headers.authorization;
+    if (!authorizationHeader) {
+      throw new ErrorWithStatusCode('Missing Authorization header', STATUS_UNAUTHORIZED);
+    }
+    return authorizationHeader.replace('Bearer ', '');
   }
 
   /**
    * Create a new API key.
    * @param {Object} requestContext - Context of the request.
-   * @returns {Promise<Response>} 201 Created with the new API key.
+   * @returns {Promise<ApiKey>} 201 Created with the new API key.
    */
   async function createApiKey(requestContext) {
     const { data, pathInfo: { headers } } = requestContext;
     const imsOrgId = headers['x-ims-gw-org-id'];
 
-    validateRequestData(data);
-    validateImsOrgId(imsOrgId);
+    try {
+      const imsUserToken = getImsUserToken(headers);
+
+      validateRequestData(data);
+      await validateImsOrgId(imsOrgId, imsUserToken);
+
+      // Check if the domains are within the limit
+      if (data.domains.length > maxDomainsPerApiKey) {
+        throw new ErrorWithStatusCode('Invalid request: Exceeds the number of domains allowed', STATUS_FORBIDDEN);
+      }
+
+      const { authInfo: { profile } } = attributes;
+
+      // Currently the email is assigned as the imsUserId
+      const imsUserId = profile.email;
+
+      // Check whether the user has already created the maximum number of
+      // active API keys for the given imsOrgId.
+      const apiKeys = await dataAccess.getApiKeysByImsUserIdAndImsOrgId(imsUserId, imsOrgId);
+
+      log.debug('Retrieved the API keys for the user: ', apiKeys);
+
+      const validApiKeys = apiKeys && apiKeys.filter(
+        (apiKey) => apiKey.isValid(),
+      );
+
+      if (validApiKeys && validApiKeys.length >= maxApiKeys) {
+        throw new ErrorWithStatusCode('Invalid request: Exceeds the number of API keys allowed', STATUS_FORBIDDEN);
+      }
+
+      // Create the API key
+      const apiKey = crypto.randomUUID();
+      const hashedApiKey = hashWithSHA256(apiKey);
 
-    if (data) {
-      const error = new ErrorWithStatusCode('Create API key not implemented', 501);
+      let scopes = {};
+
+      // We manually set the scopes initially to imports.read and imports.write
+      if (data.features.includes('imports')) {
+        scopes = [
+          {
+            name: 'imports.read',
+          },
+          {
+            name: 'imports.write',
+            domains: data.domains,
+          },
+        ];
+      }
+
+      const expiresAt = new Date();
+      expiresAt.setMonth(expiresAt.getMonth() + 6);
```

Comment:
Worth making this TTL configurable via a secret? Filing a ticket for the backlog is fine, too.

Reply:
JIRA: https://jira.corp.adobe.com/browse/SITES-26990